src/shared/clients/s3Client.ts

```diff
@@ -0,0 +1,94 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+
+export const DEFAULT_MAX_KEYS = 300
+export const DEFAULT_DELIMITER = '/'
+
+export interface S3Client {
+    /**
+     * Creates a bucket in the region of the client.
+     */
+    createBucket(request: CreateBucketRequest): Promise<CreateBucketResponse>
+
+    listBuckets(): Promise<ListBucketsResponse>
+
+    listObjects(request: ListObjectsRequest): Promise<ListObjectsResponse>
+
+    createFolder(request: CreateFolderRequest): Promise<CreateFolderResponse>
+
+    downloadFile(request: DownloadFileRequest): Promise<void>
+    uploadFile(request: UploadFileRequest): Promise<void>
+}
+
+export interface Bucket {
+    readonly name: string
+    readonly region: string
+    readonly arn: string
+}
+
+export interface Folder {
+    readonly name: string
+    readonly path: string
+    readonly arn: string
+}
+
+export interface File {
+    readonly name: string
+    readonly key: string
+    readonly arn: string
+    readonly lastModified?: Date
+    readonly sizeBytes?: number
```

Comment:
When would these be undefined?

Reply:
Great question. No idea.

The spec /sdk lists it as on optional field on an object. Possibly because object is reused as an input type?

https://docs.aws.amazon.com/AmazonS3/latest/API/API_Object.html#AmazonS3-Type-Object-LastModified

Rather than risk it, I just kept it optional.

Reply:
However, "Key" is also optional, and I thought that was worth risking 😄. Thoughts?